File: IcalParser.cpp
using namespace std;
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include "string.h"

#include "IcalEvent.h"

struct dateTime{
    string date;
    string time;
};

/*
    Trims whitespace from beginning and end of string s
*/
void TrimWhiteSpace(string& s){
    s.erase(0, s.find_first_not_of(" \t\r\n"));
    s.erase(s.find_last_not_of(" \t\r\n") + 1);
}

/*
Removes duplicate words from string s
*/
void RemoveDuplicateWords(string& s) {
    vector<string> words;
    string readWord;
    string newString;
    bool duplicateFlag = false;

    TrimWhiteSpace(s);

    stringstream stream(s);

    while(stream >> readWord){
        for(string i : words){
            if(i == readWord){
                duplicateFlag = true;
            }
        }
        if(!duplicateFlag){
            words.push_back(readWord);
        }
    }

    for(string i : words){
        newString.append(i);
        newString.append(" ");
    }
    s = newString;
    TrimWhiteSpace(s);
}

string StartDateTrimmer(const string& input){
    string newString = input.substr(strlen("DTSTART:"), input.length() - strlen("00Z"));

    return newString;
}

string EndDateTrimmer(const string& input){
    string newString = input.substr(strlen("DTEND:"), input.length() - strlen("00Z"));

    return newString;
}

dateTime ParseDateTime(const string& input){
    dateTime newDateTime;

    newDateTime.date = input.substr(0, input.find("T"));
    newDateTime.time = input.substr(input.find("T")+1, input.length());

    return newDateTime;
}

int LoadFile(string fileName, vector <IcalEvent> &schedule){
    string textLine;

    ifstream icalFile(fileName);
    if(!icalFile.good()){
        cout<<"ERROR reading file"<<endl;
        return 1;
    }

    // Skip VCALENDAR beginning
    getline(icalFile, textLine);
    getline(icalFile, textLine);
    getline(icalFile, textLine);
    getline(icalFile, textLine);
    getline(icalFile, textLine);

    // Read BEGIN:VEVENT
    getline(icalFile, textLine);

    while(textLine.compare("END:VCALENDAR") != 0){
        string startDate;
        getline(icalFile, startDate);
        
        // FIXA DETTA
        dateTime start = ParseDateTime(startDate);
        startDate = StartDateTrimmer(start.date);
        string startTime = start.time;
        string startTime;
        startTime = startDate.substr(startDate.find("T")+1, startDate.length());

        startDate = startDate.substr(0, startDate.find("T"));

        string endDate;
        getline(icalFile, endDate);
        // Trim string
        endDate = EndDateTrimmer(endDate);

        cout<<startDate<<endl;
        cout<<endDate<<endl;


        string endTime;
        endTime = endDate.substr(endDate.find("T")+1, endDate.length());

        endDate = endDate.substr(0, endDate.find("T"));

        // Skip DTSTAMP, UID, CREATED, LAST-MODIFIED
        getline(icalFile, textLine);
        getline(icalFile, textLine);
        getline(icalFile, textLine);
        getline(icalFile, textLine);

        string location;
        getline(icalFile, location);
        // Trim string
        location = location.substr(9, location.length());

        // Skip SEQUENCE, STATUS
        getline(icalFile, textLine);
        getline(icalFile, textLine);

        string summary;
        getline(icalFile, summary);

        // Find program
        int programStart = summary.find("Program: ");
        if(programStart != -1) programStart += 9;
        int programEnd = summary.find("Kurs.grp: ");
        if(programEnd != -1) programEnd -= programStart;

        string program;
        if(programStart == -1 || programEnd == -1){
            program = "N/A";
        }else{
            program = summary.substr(programStart, programEnd);
        }

        // Find Kurs
        int kursStart = summary.find("Kurs.grp: ");
        if(kursStart != -1) kursStart += 10;
        int kursEnd = summary.find("Sign: ");
        if(kursEnd != -1) kursEnd -= kursStart;

        string kurs;
        if(kursStart == -1 || kursEnd == -1){
            kurs = "N/A";
        }else{
            kurs = summary.substr(kursStart, kursEnd);
        }

        // Find moment
        int momentStart = summary.find("Moment: ");
        if(momentStart != -1) momentStart += 8;
        int momentEnd = summary.find("Aktivitetstyp: ");
        if(momentEnd != -1) momentEnd -= momentStart;

        string moment;
        if(momentStart == -1 || momentEnd == -1){
            moment = "N/A";
        }else{
            moment = summary.substr(momentStart, momentEnd);
        }

        TrimWhiteSpace(startDate);
        TrimWhiteSpace(endDate);
        TrimWhiteSpace(location);
        TrimWhiteSpace(program);
        TrimWhiteSpace(kurs);
        TrimWhiteSpace(moment);

        RemoveDuplicateWords(kurs);

        IcalEvent newEvent = IcalEvent(startDate, endDate, startTime, endTime, location, program, kurs, moment);
        schedule.push_back(newEvent);

        //Skip TRANSP, X-MICROSOFT, END:VEVENT
        getline(icalFile, textLine);
        getline(icalFile, textLine);
        getline(icalFile, textLine);

        // Read next BEGIN:VEVENT or END:VCALENDAR
        getline(icalFile, textLine);
        // Trim for trailing whitespace/special characters
        TrimWhiteSpace(textLine);
        //textLine.erase(0, textLine.find_first_not_of(" \t\r\n"));
        //textLine.erase(textLine.find_last_not_of(" \t\r\n") + 1);
    }

    icalFile.close();
    return 0;
}

int main(){
    vector <IcalEvent> schedule;
    LoadFile("SchemaICAL.ics", schedule);

    /*
    for(IcalEvent i : schedule){
        i.Print();
        cout<<endl;
    }
    */
}

File: webserver/icalParser.ts
import IcalEvent from "./IcalEvent.js";

function extractDate(date: string):string{
  let parsedDate: string = "N/A";
  
  let startIndex: number = 0;
  
  if(date.includes("DTSTART")){
    startIndex = "DTSTART:".length;
  }else if(date.includes("DTEND:")){
    startIndex = "DTEND:".length;
  }else{
    return "Error parsing date";
  }
  
  let endIndex: number = date.lastIndexOf("T");
  
  if(startIndex >= endIndex){
    return "Error parsing date"
  }else{
    parsedDate = date.slice(startIndex, endIndex);
  }
  
  return parsedDate;
}

function extractLocation(location: string):string{
  let parsedLocation: string = "N/A";
  
  let startIndex: number = "LOCATION:".length;
  
  parsedLocation = location.slice(startIndex);
  
  return parsedLocation;
}

function extractTime(date: string):string {
  let time: string = "N/A";
  
  let tPos: number = date.lastIndexOf("T");
  if(tPos === -1){
    return "Error extracting time";
  }else{
    time = date.slice(tPos+1, date.length);
    time = time.slice(0, time.indexOf("00Z"));
  }
  
  return time;
}

function parseSummary(summary: string):string[]{
  const arr: string[] = ["N/A2", "N/A2", "N/A2"];
  const summaryStruct = {
    program: "NO PROGRAM",
    kurs: "NO COURSE",
    moment: "NO MOMENT"
  };
  
//Program: Kurs.grp: Sign: Moment: Aktivitetstyp:

  let programStart: number = 0,
  programEnd: number = 0,
  kursStart: number = 0,
  kursEnd: number = 0,
  momentStart: number = 0,
  momentEnd: number = 0;

  const programExists: boolean = summary.includes("Program: ");
  const kursExists: boolean = summary.includes("Kurs.grp: ");
  const signExists: boolean = summary.includes("Sign: ");
  const momentExists: boolean = summary.includes("Moment: ");
  const aktivitetstypExists: boolean = summary.includes("Aktivitetstyp: ");
  
  if(programExists){
    programStart = (summary.indexOf("Program: "))+("Program: ".length);
    
    if(kursExists){
      programEnd = summary.indexOf("Kurs.grp: ")-1;
    }else if(signExists){
      programEnd = summary.indexOf("Sign: ")-1;
    }else if(momentExists){
      programEnd = summary.indexOf("Moment: ")-1;
    }else if(aktivitetstypExists){
      programEnd = summary.indexOf("Aktivitetstyp: ")-1;
    }
    
    arr[0] = summary.slice(programStart, programEnd);
  }
  
  if(kursExists){
    kursStart = (summary.indexOf("Kurs.grp: "))+("Kurs.grp: ".length);

    if(signExists){
      kursEnd = summary.indexOf("Sign: ")-1;
    }else if(momentExists){
      kursEnd = summary.indexOf("Moment: ")-1;
    }else if(aktivitetstypExists){
      kursEnd = summary.indexOf("Aktivitetstyp: ")-1;
    }else{
      kursEnd = summary.length -1;
    }

    arr[1] = summary.slice(kursStart, kursEnd);
  }

  if(momentExists){
    momentStart = (summary.indexOf("Moment: "))+("Moment: ".length);

    if(aktivitetstypExists){
      momentEnd = summary.indexOf("Aktivitetstyp: ")-1;
    }else{
      momentEnd = summary.length -1;
    }
    
    arr[2] = summary.slice(momentStart, momentEnd);
  }
  
  
  return arr;
}

function removeDuplicateWords(s: string):string{
  let newWord: string = "";
  const arr: string[] = s.split(" ");
  const words = new Set<string>;
  
  for(let word of arr){
    //console.log(word);
    words.add(word);
  }
  
  for(let word of words){
    //console.log(word);
    newWord = newWord.concat(word);
    newWord = newWord.concat(" ");
  }
  
  newWord = newWord.trimEnd();
  
  //console.log("newword: "+newWord);
  
  return newWord;
}

export default function parseIcalToJson(icalFile: string): IcalEvent[]{
  let startDate: string = "N/A",
  endDate: string = "N/A",
  startTime: string = "N/A",
  endTime: string = "N/A",
  location: string = "N/A",
  summary: string = "N/A",
  program: string = "N/A",
  kurs: string = "N/A",
  moment: string = "N/A";
  
  const events: IcalEvent[] = [];

  /*
  if(icalFile !== 'string'){
    console.log("WTF");
  }
  */
  
  const icalFileLines = icalFile.split(/\r?\n/);
  
  for(let line of icalFileLines){
    if(line.startsWith("DTSTART")){
      startDate = line;
    }else if(line.startsWith("DTEND")){
      endDate = line;
    }else if(line.startsWith("LOCATION:")){
      location = line;
    }else if(line.startsWith("SUMMARY")){
      summary = line;
    }else if(line.includes("END:VEVENT")){
      startTime = extractTime(startDate);
      endTime = extractTime(endDate);
      
      startDate = extractDate(startDate);
      endDate = extractDate(endDate);
      
      location = extractLocation(location);
      
      let summaryArr: string[] = parseSummary(summary);
      
      program = summaryArr[0];
      
      kurs = summaryArr[1];
      kurs = removeDuplicateWords(kurs);
      
      moment = summaryArr[2];
      
      const newEvent = new IcalEvent(startDate, endDate, startTime, endTime, location, program, kurs, moment);
      
      events.push(newEvent);
    }
  }
  return events;
}